src/GitHub/App.hs

```diff
@@ -0,0 +1,152 @@
+authenticateInstallation :: InstallationAuth -> IO Auth
+authenticateInstallation instAuth = do 
+    gtaToken    <- readIORef (instAuth ^. token)
+    currentTime <- getPOSIXTime
+    if (expireAt gtaToken) - currentTime >= bufferTime 
+    then return $ OAuth $ encodeUtf8 $ instToken gtaToken
+    else do 
+        let applicationId = instAuth ^. appId
+        OAuth <$> encodeUtf8 <$> instToken <$> (join $ readIORef <$> renewInstAuthToken instAuth) 
+
+
+-- | Gets new token from github and writes it to IORef in given InstallationAuth. Also returns that token. 
+renewInstAuthToken :: InstallationAuth -> IO (IORef GTAToken) 
+renewInstAuthToken instAuth = do 
+    time    <- getPOSIXTime
+    let tkn = instAuth ^. token
+    case (makeJWT time (instAuth ^. appId) (instAuth ^. appPrivateKey)) of 
+        Left  err  -> throwM (FailedGenerateJWT err) 
+        Right jwt  -> do 
+            t <- request (https baseURL /: "installations" /: (instAuth ^. installationId) /: "access_tokens") mempty jwt    
+            writeIORef tkn t 
+            return tkn
+
+-- | Creates new InstallationAuth value. Useing this function is the only way to create InstallationAuth value, 
+-- because we don't export constructors. 
+createInstAuth :: Int -> PrivateKey -> InstallationId -> IO InstallationAuth 
+createInstAuth appId key instId = do 
+    time    <- getPOSIXTime
+    case (makeJWT time appId key) of 
+        Left  err  -> throwM (FailedGenerateJWT err) 
+        Right jwt  -> do 
+            t      <- request (https baseURL /: "installations" /: instId /: "access_tokens") mempty jwt
+            token  <- newIORef t
+            return $ InstallationAuth appId key instId token
+    
+
+-- | Creates JSON Web Token for given application Id using application's privateKey. 
+makeJWT :: POSIXTime -> Int -> PrivateKey -> Either Text JSON 
+makeJWT currentTime appId appPrivateKey = do 
+    let currDate = numericDate currentTime 
+    let expDate  = numericDate (currentTime + jwtExpTime)
+    if (currDate >> expDate) == Nothing 
+    then Left "Failed to convert time to numericDate."
+    else do
+        let issuer = stringOrURI $ show appId 
+        if issuer == Nothing 
+        then Left $ "Failed to convert id " <> show appId <> " to StringOrURI type"
+        else do 
+            let jwtClaimsSet = mempty {iss = issuer, iat = currDate, exp = expDate}
+            return $ encodeSigned (RSAPrivateKey appPrivateKey) jwtClaimsSet
+
+-- | Make request to github to get installation Auth token. 
+request :: FromJSON m => Url scheme -> Option scheme -> JSON -> IO m
+request url opts jwt = runReq def $ responseBody <$>
+  req POST url NoReqBody
+    jsonResponse -- specify how to interpret response
+     (  header "Authorization" ("Bearer " <> encodeUtf8 jwt)
+     <> header "Accept" "application/vnd.github.machine-man-preview+json"
+     <> header "user-agent" "req"
```

Comment:
Please, something that makes sense.

Reply:
Use "Haskell req-1.2.1" as user agent now.